src/plonk/permutation/prover.rs

```diff
@@ -0,0 +1,391 @@
+                                    .iter_mut()
+                                    .zip(advice[start..].iter())
+                                    .zip(permutation[start..].iter())
+                                {
+                                    *left *= &(*advice + &(*beta * permutation) + &gamma);
+                                }
+                            });
+                        }
+
+                        let mut right = cosets_inv;
+                        let mut current_delta = *beta * &C::Scalar::ZETA;
+                        let step = domain.get_extended_omega();
+                        for advice in p.columns.iter().map(|&column| {
+                            &advice_cosets[pk.vk.cs.get_advice_query_index(column, 0)]
+                        }) {
+                            parallelize(&mut right, move |right, start| {
+                                let mut beta_term =
+                                    current_delta * &step.pow_vartime(&[start as u64, 0, 0, 0]);
+                                for (right, advice) in right.iter_mut().zip(advice[start..].iter())
+                                {
+                                    *right *= &(*advice + &beta_term + &gamma);
+                                    beta_term *= &step;
+                                }
+                            });
+                            current_delta *= &C::Scalar::DELTA;
+                        }
+
+                        left - &right
+                    }),
+            );
+
+        Ok((
+            Constructed {
+                permutation_product_polys: self.permutation_product_polys,
+                permutation_product_blinds: self.permutation_product_blinds,
+                permutation_product_commitments: self.permutation_product_commitments,
+            },
+            expressions,
+        ))
+    }
+}
+
+impl<C: CurveAffine> super::ProvingKey<C> {
+    fn evaluate(&self, x: ChallengeX<C::Scalar>) -> Vec<C::Scalar> {
+        self.polys
+            .iter()
+            .map(|poly| eval_polynomial(poly, *x))
+            .collect()
+    }
+
+    fn open<'a>(
+        &'a self,
+        evals: &'a [C::Scalar],
+        x: ChallengeX<C::Scalar>,
+    ) -> impl Iterator<Item = ProverQuery<'a, C>> + Clone {
+        self.polys
+            .iter()
+            .zip(evals.iter())
+            .map(move |(poly, eval)| ProverQuery {
+                point: *x,
+                poly,
+                blind: Blind::default(),
+                eval: *eval,
+            })
+    }
+}
+
+impl<C: CurveAffine> Constructed<C> {
+    pub(crate) fn evaluate<HBase: Hasher<C::Base>, HScalar: Hasher<C::Scalar>>(
+        self,
+        pk: &ProvingKey<C>,
+        x: ChallengeX<C::Scalar>,
+        transcript: &mut Transcript<C, HBase, HScalar>,
+    ) -> Evaluated<C> {
+        let domain = &pk.vk.domain;
+
+        let permutation_product_evals: Vec<C::Scalar> = self
```

Comment:
They need `Vec<_>` at a minimum, because we then need to iterate over them to add them to the transcript (and the type needs to be known at that point). But the `C::Scalar` is not necessary here.

Reply:
ah. that does seems like a defect in the rust compiler's inference capabilities though